scripts/backport.py

```diff
@@ -0,0 +1,371 @@
+            f"{commit_sha[:9]}\tskip\talready in branch\t{commit_sha[:9]}\t{commit_title}"
+        )
+        continue
+
+    pygithub_commit = source_repo.get_commit(sha=commit_sha)
+
+    pulls = pygithub_commit.get_pulls()
+    if not pulls:
+        print(
+            f"{commit_sha[:9]}\terror\tlinked pr not found\t{commit_sha[:9]}\t{commit_title}"
+        )
+        continue
+
+    pull = pulls[0]
+    issue_number = get_referenced_issue(pull.number)
+    if not issue_number:
+        print(
+            f"{commit_sha[:9]}\tskip\tdoes not close an issue\t{pull.number}\t{pull.title}"
+        )
+        continue
+
+    pr_labels = {label.name for label in pull.labels}
+    stopper_labels = pr_labels.intersection(["no-backport", "pr-auto-backport-failed"])
+    if stopper_labels:
+        print(
+            f"{commit_sha[:9]}\tskip\tpull request labeled as '{list(stopper_labels)[0]}'\t{pull.number}\t{pull.title}"
+        )
+        continue
+
+    changed_files = {file.filename for file in pull.get_files()}
+    stopper_files = changed_files.intersection(
+        ["sql/updates/latest-dev.sql", "sql/updates/reverse-dev.sql"]
+    )
+    if stopper_files:
+        print(
+            f"{commit_sha[:9]}\tskip\tpull request modifies '{list(stopper_files)[0]}'\t{pull.number}\t{pull.title}"
+        )
+        continue
+
+    issue = source_repo.get_issue(number=issue_number)
+    issue_labels = {label.name for label in issue.labels}
+
+    if "bug" not in issue_labels:
+        print(
+            f"{commit_sha[:9]}\tskip\tissue not marked as 'bug'\t{issue.number}\t{issue.title}"
+        )
+        continue
+
+    if "no-backport" in issue_labels:
+        print(
+            f"{commit_sha[:9]}\tskip\tissue marked as 'no-backport'\t{issue.number}\t{issue.title}"
+        )
+        continue
+
+    print(f"{commit_sha[:9]}\tbackport\t\t{commit_sha}\t{commit_title}")
+
+    # Remember the PR and the corresponding resulting commit in main.
+    if pull.number not in prs_to_backport:
+        prs_to_backport[pull.number] = PRInfo(pull, issue_number)
+
+    # We're traversing the history backwards, and want to have the list of
+    # commits in forward order.
+    prs_to_backport[pull.number].pygithub_commits.insert(0, pygithub_commit)
+
+
+# Now, go over the list of PRs that we have collected, and try to backport
+# each of them.
+for pr_info in prs_to_backport.values():
+    original_pr = pr_info.pygithub_pr
+    backport_branch = f"backport/{backport_target}/{original_pr.number}"
+
+    # If there is already a backport branch for this PR, just skip it.
+    if (
+        subprocess.run(
+            f"git rev-parse {target_remote}/{backport_branch} > /dev/null 2>&1",
+            shell=True,
+            check=False,
+        ).returncode
+        == 0
+    ):
+        print(
+            f'Backport branch {backport_branch} for PR #{original_pr.number}: "{original_pr.title}" already exists. Skipping.'
+        )
+        continue
```

Comment:
It is a lot easier to read if you split out the git commands into separate functions and use them in the code.

Reply:
I don't think so... the first thing I need to know when I look at this code is which git command it is running. Putting this command somewhere else just makes it more complicated. I agree these lines look kinda ugly, especially when formatted by "black", but it's all just boilerplate around the only meaningful thing, the git command.

I can instead move boilerplate to functions, e.g. `git_check_output` and `git_returncode`, to remove the amount of visual noise. Thankfully I only need two these modes for subprocess.

Reply:
> Thankfully I only need two these modes for subprocess.

Actually three, but anyway, done that, looks better now.